os/drivers/compression/compress.c

```diff
@@ -116,10 +130,52 @@ static int comp_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
 		ret = OK;
 		break;
  	case COMPIOC_DECOMPRESS:
+		if(comp_info == NULL) {
+			return -EINVAL;
+		}
 		ret = decompress_block(comp_info->output_buffer, &comp_info->output_size, comp_info->input_buffer, &comp_info->input_size);
 		if (ret == ENOMEM) {
 			bcmpdbg("Output buffer allocated is not sufficient\n");
 		}
+#ifdef CONFIG_COMPRESSED_BINARY
+	case COMPIOC_FDCMP_INIT:
+		if((char *)arg == NULL) {
+			return -EINVAL;
+		}
+		data = (struct file_decomp_data_s *)malloc(sizeof(struct file_decomp_data_s));
+		data->fd = open((char *)arg, "r");
+		filep->f_priv = data;
+		if (data->fd < 0) {
+			berr("Failed to open file = %s\n", (char *)arg);
+			return -EINVAL;
+		}
+		ret = compress_init(data->fd, 0, &filelen);
+		break;
+	case COMPIOC_FDCMP_GET_BUFSIZE:
+		data = filep->f_priv;
+		data->compression_header = get_compression_header();
+		ret = data->compression_header->binary_size;
+		break;
+	case COMPIOC_FDCMP_DECOMPRESS:
+		if((uint8_t *)arg == NULL) {
+			return -EINVAL;
+		}
+		data = filep->f_priv;
+		size = compress_read(data->fd, 0, (uint8_t *)arg, data->compression_header->binary_size, 0);
```

Comment:
Need to check if compress read can return an error. If so, we cannot assume that it is the size. 
We will need to handle the error code and free the resources.